wordpress-mobile/WordPress-iOS
# PingHub Manager, or how to stay connected

Keep PingHub connected, as long as possible and if it makes sense.

I've tried to add relevant explanations across the code but I'll repeat some here.

PingHubManager will attempt to keep a PinghubClient connected as long as it is alive while certain conditions are met.

## When it connects

The manager will try to connect as long as it has an oAuth2 token and the app is in the foreground.
When a connection fails for some reason, it will try to reconnect whenever there is an internet connection, as detected by Reachability. If the app thinks it’s online but connections are still failing, the manager adds an increasing delay to the reconnection to avoid too many attempts.

## Debugging

There are a couple helpers to aid with debugging the manager.
First, if you want to see a detailed log of every state changed and the resulting action in the console, add -debugPinghub to the arguments passed on launch in Xcode’s scheme editor. I've left the argument present but disabled in the scheme, so you only have to enable it.

![screen shot 2016-12-22 at 19 04 12](https://cloud.githubusercontent.com/assets/8739/21435313/7b1c1268-c879-11e6-932b-1b02cc33d6f5.png)

Second, if you want to simulate some network error conditions to test the retry algorithm, you can set a breakpoint in any Swift code, and enter into the LLDB prompt:
expr PinghubClient.Debug.simulateUnreachableHost(true)
This will simulate a “Unreachable Host” error every time the PingHub client tries to connect. If you want to disable it, do the same thing but passing false.

## To test

- Launch the app with debugging enabled and look at the console, see if PingHub has connected.
- Change the state into all the combinations you can think of (log in/out, foreground/background, internet connection on/off, simulate/real network errors)
- Verify that PingHub reconnects as expected

See also: #6347 #6164
Needs review: @jleandroperez
